File: frontend/src/types/restoration.ts
import type { AnnotationCoordinates } from './annotation';
import type { MuralImage } from './mural';

export type RestorationMode = 'full' | 'partial';
export type RestorationRunStatus = 'succeeded' | 'failed';
export type RestorationSourceType = 'primary' | 'variant';

export interface RestorationParameters {
  restorationStrength: number;
  cleaningLevel: number;
  colorRecovery: number;
  detailPreservation: number;
  crackRepairBias: number;
  structureClosure: number;
  structureFill: number;
  edgeBlend: number;
  stainRemoval: number;
  moldSuppression: number;
  saltReduction: number;
  toneCorrection: number;
  localColorRepair: number;
  textureRebuild: number;
  outputPreference: 'clarity' | 'fidelity';
  randomness: number;
}

export interface RestorationResult {
  id: string;
  runId?: string;
  parentResultId?: string | null;
  imagePath?: string;
  imageUrl: string;
  imageHash?: string;
  width?: number;
  height?: number;
  fileSize?: number;
  isMock: boolean;
  providerName?: string;
  createdAt: string;
  parametersSnapshot: RestorationParameters;
  sourceType: RestorationSourceType;
  committedMuralImageId?: string | null;
}

export interface RestorationRun {
  id: string;
  muralId: string;
  mode: RestorationMode;
  sourceImagePath: string;
  sourceImageUrl: string;
  sourceImageHash: string;
  sourceImageWidth: number;
  sourceImageHeight: number;
  sourceImageSize: number;
  parametersSnapshot: RestorationParameters;
  annotationIds: string[];
  manualSelection: AnnotationCoordinates | null;
  status: RestorationRunStatus;
  latestResultId?: string | null;
  committedResultId?: string | null;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

export interface RestorationRunDetail {
  run: RestorationRun;
  results: RestorationResult[];
  currentResult: RestorationResult | null;
  variants: RestorationResult[];
  sourceImageUrl: string;
}

export interface RestorationCommitResponse {
  run: RestorationRun;
  result: RestorationResult;
  muralImage: MuralImage;
}

export interface RestorationPreflightInput {
  muralId?: string;
  hasSourceImage: boolean;
  mode: RestorationMode;
  annotationIds: string[];
  manualSelection: AnnotationCoordinates | null;
}

File: backend/internal/model/restoration.go
package model

import (
	"time"

	"gorm.io/datatypes"
)

type RestorationMode string

const (
	RestorationModeFull    RestorationMode = "full"
	RestorationModePartial RestorationMode = "partial"
)

type RestorationRunStatus string

const (
	RestorationRunSucceeded RestorationRunStatus = "succeeded"
	RestorationRunFailed    RestorationRunStatus = "failed"
)

type RestorationSourceType string

const (
	RestorationSourcePrimary RestorationSourceType = "primary"
	RestorationSourceVariant RestorationSourceType = "variant"
)

type RestorationRun struct {
	ID                 string               `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	MuralID            string               `gorm:"type:uuid;not null;index" json:"muralId"`
	Mode               RestorationMode      `gorm:"type:varchar(20);not null" json:"mode"`
	SourceImagePath    string               `gorm:"not null" json:"sourceImagePath"`
	SourceImageHash    string               `gorm:"not null" json:"sourceImageHash"`
	SourceImageWidth   int                  `json:"sourceImageWidth"`
	SourceImageHeight  int                  `json:"sourceImageHeight"`
	SourceImageSize    int64                `json:"sourceImageSize"`
	ParametersSnapshot datatypes.JSON       `gorm:"type:jsonb;not null;default:'{}'" json:"parametersSnapshot"`
	AnnotationIDs      datatypes.JSON       `gorm:"type:jsonb;not null;default:'[]'" json:"annotationIds"`
	ManualSelection    datatypes.JSON       `gorm:"type:jsonb" json:"manualSelection"`
	Status             RestorationRunStatus `gorm:"type:varchar(20);not null" json:"status"`
	LatestResultID     *string              `gorm:"type:uuid" json:"latestResultId"`
	CommittedResultID  *string              `gorm:"type:uuid" json:"committedResultId"`
	CreatedBy          string               `gorm:"not null" json:"createdBy"`
	CreatedAt          time.Time            `json:"createdAt"`
	UpdatedAt          time.Time            `json:"updatedAt"`

	Results []RestorationResult `gorm:"foreignKey:RunID" json:"results,omitempty"`
}

type RestorationResult struct {
	ID                    string                `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	RunID                 string                `gorm:"type:uuid;not null;index" json:"runId"`
	ParentResultID        *string               `gorm:"type:uuid;index" json:"parentResultId"`
	SourceType            RestorationSourceType `gorm:"type:varchar(20);not null" json:"sourceType"`
	ImagePath             string                `gorm:"not null" json:"imagePath"`
	ImageHash             string                `gorm:"not null" json:"imageHash"`
	Width                 int                   `json:"width"`
	Height                int                   `json:"height"`
	FileSize              int64                 `json:"fileSize"`
	ParametersSnapshot    datatypes.JSON        `gorm:"type:jsonb;not null;default:'{}'" json:"parametersSnapshot"`
	IsMock                bool                  `json:"isMock"`
	ProviderName          string                `gorm:"not null" json:"providerName"`
	CommittedMuralImageID *string               `gorm:"type:uuid" json:"committedMuralImageId"`
	CreatedAt             time.Time             `json:"createdAt"`
}
